File: DentalWave/src/main/java/com/summerproject2026/DentalWave/service/EmployeeService.java
package com.summerproject2026.DentalWave.service;

import com.summerproject2026.DentalWave.dto.AvailabilityDto;
import com.summerproject2026.DentalWave.dto.EmployeeDto;
import com.summerproject2026.DentalWave.enums.WorkStatus;

import java.util.List;

/**
 * Service interface defining all business operations for Employee management.
 *
 * Covers:
 *  - Full CRUD for employees
 *  - Filtering by office and work status
 *  - Nested availability management (add / update / delete)
 */
public interface EmployeeService {

    /**
     * Creates and persists a new employee.
     * The linked User must already exist; userId in the DTO is used to resolve it.
     */
    EmployeeDto createEmployee(EmployeeDto employeeDto);

    /**
     * Retrieves an employee by primary key.
     *
     * @throws com.summerproject2026.DentalWave.exception.ResourceNotFoundException if not found
     */
    EmployeeDto getEmployeeById(Long id);

    /** Returns all employees in the system (may be empty). */
    List<EmployeeDto> getAllEmployees();

    /**
     * Updates an existing employee's fields.
     * The linked User account is NOT changed on update.
     *
     * @throws com.summerproject2026.DentalWave.exception.ResourceNotFoundException if not found
     */
    EmployeeDto updateEmployee(Long id, EmployeeDto employeeDto);

    /**
     * Deletes an employee and all cascaded data (availabilities, time-off requests).
     *
     * @throws com.summerproject2026.DentalWave.exception.ResourceNotFoundException if not found
     */
    void deleteEmployee(Long id);

    /**
     * Returns all employees assigned to a given office.
     */
    List<EmployeeDto> getEmployeesByOffice(Long officeId);

    /**
     * Returns all employees with the given work status (e.g. ACTIVE, ON_LEAVE).
     */
    List<EmployeeDto> getEmployeesByStatus(WorkStatus status);

    /**
     * Adds a new availability record to an employee's schedule.
     *
     * @throws com.summerproject2026.DentalWave.exception.ResourceNotFoundException if employee not found
     */
    AvailabilityDto addAvailability(Long employeeId, AvailabilityDto availabilityDto);

    /**
     * Updates an existing availability record that belongs to the given employee.
     *
     * @throws com.summerproject2026.DentalWave.exception.ResourceNotFoundException if either not found
     * @throws IllegalArgumentException if the availability does not belong to this employee
     */
    AvailabilityDto updateAvailability(Long employeeId, Long availabilityId,
                                       AvailabilityDto availabilityDto);

    /**
     * Deletes an availability record from an employee.
     *
     * @throws com.summerproject2026.DentalWave.exception.ResourceNotFoundException if either not found
     * @throws IllegalArgumentException if the availability does not belong to this employee
     */
    void deleteAvailability(Long employeeId, Long availabilityId);
}

File: DentalWave-frontend/src/services/EmployeeService.js
import axios from 'axios'
import { getAuthHeader } from './AuthService'

/**
 * Base URL for Employee endpoints.
 */
const EMPLOYEE_REST_API_BASE_URL = 'http://localhost:8080/api/employees'

/**
 * Creates a new employee.
 *
 * @param employee EmployeeDto
 * @returns Created EmployeeDto
 */
export const createEmployee = (employee) =>
    axios.post(EMPLOYEE_REST_API_BASE_URL, employee, getAuthHeader())

/**
 * Retrieves an employee by id.
 *
 * @param id Employee id
 * @returns EmployeeDto
 */
export const getEmployeeById = (id) =>
    axios.get(`${EMPLOYEE_REST_API_BASE_URL}/${id}`, getAuthHeader())

/**
 * Retrieves all employees.
 *
 * @returns List<EmployeeDto>
 */
export const getAllEmployees = () =>
    axios.get(EMPLOYEE_REST_API_BASE_URL, getAuthHeader())

/**
 * Updates an employee.
 *
 * @param id Employee id
 * @param employee Updated EmployeeDto
 * @returns Updated EmployeeDto
 */
export const updateEmployee = (id, employee) =>
    axios.put(`${EMPLOYEE_REST_API_BASE_URL}/${id}`, employee, getAuthHeader())

/**
 * Deletes an employee.
 *
 * @param id Employee id
 */
export const deleteEmployee = (id) =>
    axios.delete(`${EMPLOYEE_REST_API_BASE_URL}/${id}`, getAuthHeader())

/**
 * Retrieves employees by office id.
 *
 * @param officeId Office id
 * @returns List<EmployeeDto>
 */
export const getEmployeesByOffice = (officeId) =>
    axios.get(`${EMPLOYEE_REST_API_BASE_URL}/office/${officeId}`, getAuthHeader())

/**
 * Retrieves employees by work status.
 *
 * Example:
 * getEmployeesByStatus("ACTIVE")
 *
 * @param status WorkStatus value
 * @returns List<EmployeeDto>
 */
export const getEmployeesByStatus = (status) =>
    axios.get(`${EMPLOYEE_REST_API_BASE_URL}/status/${status}`, getAuthHeader())

/**
 * Searches employees by keyword.
 *
 * Example:
 * searchEmployees("jane")
 *
 * @param keyword Search term
 * @returns List<EmployeeDto>
 */
export const searchEmployees = (keyword) =>
    axios.get(
        `${EMPLOYEE_REST_API_BASE_URL}/search?keyword=${keyword}`,
        getAuthHeader()
    )

/**
 * Adds availability to an employee.
 *
 * @param employeeId Employee id
 * @param availability AvailabilityDto
 * @returns Created AvailabilityDto
 */
export const addAvailability = (employeeId, availability) =>
    axios.post(
        `${EMPLOYEE_REST_API_BASE_URL}/${employeeId}/availability`,
        availability,
        getAuthHeader()
    )

/**
 * Updates an employee availability record.
 *
 * @param employeeId Employee id
 * @param availabilityId Availability id
 * @param availability Updated AvailabilityDto
 * @returns Updated AvailabilityDto
 */
export const updateAvailability = (employeeId, availabilityId, availability) =>
    axios.put(
        `${EMPLOYEE_REST_API_BASE_URL}/${employeeId}/availability/${availabilityId}`,
        availability,
        getAuthHeader()
    )

/**
 * Deletes an employee availability record.
 *
 * @param employeeId Employee id
 * @param availabilityId Availability id
 */
export const deleteAvailability = (employeeId, availabilityId) =>
    axios.delete(
        `${EMPLOYEE_REST_API_BASE_URL}/${employeeId}/availability/${availabilityId}`,
        getAuthHeader()
    )